File: src/main/java/com/pascalnb/dbwrapper/JDBC.java
package com.pascalnb.dbwrapper;

import org.jetbrains.annotations.Contract;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * The JDBC implementation of [Database].
 */
class JDBC extends Database {

    private static final int FETCH_SIZE = 500;

    private Connection connection = null;

    @Contract("_ -> this")
    @Override
    public Database connect(boolean autoCommit) {
        if (getUrl() == null) {
            throw new DatabaseException("URL for database connection not set.");
        }
        try {
            connection = DriverManager.getConnection(getUrl(), getUsername(), getPassword());
            connection.setTransactionIsolation(Connection.TRANSACTION_SERIALIZABLE);
            connection.setAutoCommit(autoCommit);
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
        return this;
    }

    @Contract("-> this")
    @Override
    public Database connect() {
        return connect(true);
    }

    @Override
    public Database commit() {
        checkConnection();
        try {
            connection.commit();
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
        return this;
    }

    @Override
    protected void checkConnection() throws DatabaseException {
        if (connection == null) {
            throw new DatabaseException("No connection to the database exists or it has already been closed.");
        }
    }

    @Contract("_, _ -> param1")
    private PreparedStatement setVariables(PreparedStatement statement, Object... variables) {
        try {
            statement.setFetchSize(FETCH_SIZE);

            for (int i = 0; i < variables.length; i++) {
                int index = i + 1;
                Object variable = variables[i];
                if (variable == null) {
                    statement.setObject(index, null);
                    continue;
                }

                if (variable instanceof Integer) {
                    statement.setInt(index, (int) variable);
                } else if (variable instanceof Double) {
                    statement.setDouble(index, (double) variable);
                } else if (variable instanceof Boolean) {
                    statement.setBoolean(index, (boolean) variable);
                } else if (variable instanceof Long) {
                    statement.setLong(index, (long) variable);
                } else if (variable instanceof Float) {
                    statement.setFloat(index, (float) variable);
                } else if (variable instanceof Byte) {
                    statement.setByte(index, (byte) variable);
                } else if (variable instanceof Short) {
                    statement.setShort(index, (short) variable);
                } else {
                    statement.setString(index, (String) variable);
                }

            }
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
        return statement;
    }

    @Override
    public Table queryStatement(Query query) {
        checkConnection();
        try {
            return parseResult(
                setVariables(
                    connection.prepareStatement(query.toString()),
                    query.getArgs()
                ).executeQuery()
            );
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }

    @Override
    public void executeStatement(Query query) {
        checkConnection();
        try {
            setVariables(connection.prepareStatement(query.toString()), query.getArgs()).execute();
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }

    @Override
    public void close() {
        checkConnection();
        try {
            connection.close();
        } catch (SQLException e) {
            throw new DatabaseException(e);
        } finally {
            connection = null;
        }
    }

    // specific implementation to parse a ResultSet to a Table
    @Contract(value = "_ -> new")
    private Table parseResult(ResultSet resultSet) {
        try {
            ResultSetMetaData metaData = resultSet.getMetaData();
            int columnCount = metaData.getColumnCount();
            String[] attributes = new String[columnCount];

            for (int i = 0; i < attributes.length; i++) {
                attributes[i] = metaData.getColumnName(i + 1);
            }

            List<String[]> tuples = new ArrayList<>();

            while (resultSet.next()) {
                String[] tuple = new String[columnCount];

                for (int i = 0; i < columnCount; i++) {
                    tuple[i] = resultSet.getString(i + 1);
                }

                tuples.add(tuple);
            }

            return new Table(attributes, tuples);
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }

}

File: src/main/kotlin/com/pascalnb/dbwrapper/JDBC.kt
package com.pascalnb.dbwrapper

import org.jetbrains.annotations.Contract
import java.sql.*
import java.util.function.Consumer

/**
 * The JDBC implementation of [Database].
 */
@Suppress("unused")
internal class JDBC : Database() {

    private var connection: Connection? = null
    private var statement: Statement? = null

    @Contract("-> this")
    override fun connect(): Database {
        if (url == null) {
            throw DatabaseException("URL for database connection not set.")
        }
        try {
            connection = DriverManager.getConnection(url, username, password)
            connection!!.transactionIsolation = Connection.TRANSACTION_SERIALIZABLE
            connection!!.autoCommit = false
            statement = connection!!.createStatement()
        } catch (e: SQLException) {
            throw DatabaseException(e)
        }
        return this
    }

    @Throws(DatabaseException::class)
    override fun checkConnection() {
        if (connection == null || statement == null) {
            throw DatabaseException(
                "No connection to the database exists" +
                        "or it has already been closed."
            )
        }
    }

    @Contract("_ -> this")
    override fun execute(query: Query): Database {
        checkConnection()
        try {
            statement!!.execute(query.toString())
        } catch (e: SQLException) {
            throw DatabaseException(e)
        }
        return this
    }

    @Contract("_, _ -> this")
    override fun query(callback: Consumer<Table>, query: Query): Database {
        checkConnection()
        try {
            statement!!.fetchSize = FETCH_SIZE
            callback.accept(parseResult(statement!!.executeQuery(query.toString())))
        } catch (e: SQLException) {
            throw DatabaseException(e)
        }
        return this
    }

    @Contract("_, _ -> param1")
    private fun setVariables(statement: PreparedStatement, variables: Array<Any?>): PreparedStatement {
        try {
            statement.fetchSize = FETCH_SIZE

            for ((i, variable) in variables.withIndex()) {
                val index = i + 1

                if (variable == null) {
                    statement.setObject(index, null)
                    continue
                }

                when (variable) {
                    is Int -> {
                        statement.setInt(index, variable)
                    }

                    is Double -> {
                        statement.setDouble(index, variable)
                    }

                    is Boolean -> {
                        statement.setBoolean(index, variable)
                    }

                    is Long -> {
                        statement.setLong(index, variable)
                    }

                    is Float -> {
                        statement.setFloat(index, variable)
                    }

                    is Byte -> {
                        statement.setByte(index, variable)
                    }

                    is Short -> {
                        statement.setShort(index, variable)
                    }

                    else -> {
                        statement.setString(index, variable.toString())
                    }
                }
            }
        } catch (e: SQLException) {
            throw DatabaseException(e)
        }
        return statement
    }

    @Contract("_, _ -> this")
    override fun queryStatement(callback: Consumer<Table>, preparedStatement: Query): Database {
        checkConnection()
        try {
            callback.accept(
                parseResult(
                    setVariables(
                        connection!!.prepareStatement(preparedStatement.toString()),
                        preparedStatement.args
                    ).executeQuery()
                )
            )
        } catch (e: SQLException) {
            throw DatabaseException(e)
        }
        return this
    }

    @Contract("_ -> this")
    override fun executeStatement(preparedStatement: Query): Database {
        checkConnection()
        try {
            setVariables(connection!!.prepareStatement(preparedStatement.toString()), preparedStatement.args).execute()
        } catch (e: SQLException) {
            throw DatabaseException(e)
        }
        return this
    }

    @Contract("_, _ -> this")
    fun executeBatch(preparedStatement: String?, vararg variablesArray: Array<Any?>): Database {
        checkConnection()
        try {
            val statement = connection!!.prepareStatement(preparedStatement)
            for (variables in variablesArray) {
                setVariables(statement, variables).addBatch()
            }
            statement.executeBatch()
        } catch (e: SQLException) {
            throw DatabaseException(e)
        }
        return this
    }

    override fun close() {
        checkConnection()
        try {
            connection!!.commit()
            connection!!.autoCommit = true
            connection!!.close()
        } catch (e: SQLException) {
            e.printStackTrace()
        }
        connection = null
        statement = null
    }

    companion object {
        private const val FETCH_SIZE = 500

        // specific implementation to parse a ResultSet to a Table
        @Contract(value = "_ -> new")
        private fun parseResult(resultSet: ResultSet): Table {
            return try {
                val metaData = resultSet.metaData
                val columnCount = metaData.columnCount
                val attributes = Array(columnCount) { "" }

                for (i in 0 until columnCount) {
                    attributes[i] = metaData.getColumnName(i + 1)
                }

                val tuples: MutableList<Array<String?>> = ArrayList()

                while (resultSet.next()) {
                    val tuple = arrayOfNulls<String>(columnCount)

                    for (i in 0 until columnCount) {
                        tuple[i] = resultSet.getString(i + 1)
                    }

                    tuples.add(tuple)
                }

                Table(attributes, tuples)
            } catch (e: SQLException) {
                throw DatabaseException(e)
            }
        }
    }
}